### DISCUSSION TITLE
Are fields of type "id" auto-generated?

### DISCUSSION THREAD

**Discussion by ansFourtyTwo** [OP-TED/eForms-SDK#669] [Q&A] [answered]
I am implementing forms, which allow the user to enter notices. However, I stumbled over fields of type `id` and of type `id-ref`. As far as I understand, for fields of type `id-ref` the user should be able to select IDs, which were entered in some other fields of type `id`, if their scheme (`idScheme`) matches. 

For example if a field of type `id-ref` has `"idSchemes" : ["LOT", "GLO"]`, all values of input fields of type `id` with `"idScheme" : "LOT"` or `"idScheme" : "GLO"` of that notice should be selectable.

But where are those IDs generated? It seems not like the user has to enter those IDs, as at least some of the fields are hidden fields. 

Am I right, that fields of type `id` are auto-generated? Is there any restriction or is this assumption applicable to all fields of type `id`?

**Comment by YvesJo** [comment]
Hi,
You're understanding is correct, it is certainly better to have the IDs generated automatically, else the user would need to know precisely which scheme to use for a given object and would have to remember which ones have already been used to prevent duplicates. For references, your analysis is also correct. Some identifier are limited to a specific string instead of a pattern (e.g. BT-09(a)-Procedure) and should only be generated when further information about the object is provided (in the case of BT-09(a)-Procedure, description of the Cross Border Law should be provided, else the ID should not be generated)

**Comment by rouschr** [comment, marked as answer]
Hello, you can have a look at the [editor demo project](https://github.com/OP-TED/eforms-notice-editor).
The back-end is in Java and the front-end is plain JavaScript, HTML and CSS.
It demonstrates how to build forms (most of it), we demo the generation of type "id" and "id-ref".
"id-ref" is indeed some kind of select in the UI and "id" is often generated and hidden.
https://github.com/OP-TED/eforms-notice-editor/blob/develop/src/main/resources/static/js/custom/notice-type-definition.js
```
  case 'id': return new IdInputField(content, level);
  case 'id-ref': return new IdRefInputField(content, level);
```
And in notice-form.js we have the function: idFromFieldIdAndInstanceNumber
